db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -1140,11 +1176,22 @@ private void orderUsefulPredicates(Optimizable optTable,
         if (inlistQualified) {
           if (inlistPreds.size() > 1) {
             predsForNewInList = new ArrayList<>();
-            int firstPred = -1, lastPred = -1, lastIndexPos = -1, firstInListPred = -1;
+            int firstPred = -1, lastPred = -1, lastIndexPos = -1 + firstColumnIdx, firstInListPred = -1;
 
             boolean foundPred[] = new boolean[usefulCount];
-            int numConstants = 1;
+            boolean indexPrefixMissing = accessPath.getNumUnusedLeadingIndexFields() > 0;
+            int firstColumnIteratedValues = !indexPrefixMissing ? 1 :
+                          (int) accessPath.getConglomerateDescriptor().
+                                           getFirstColumnStats().
+                                           getFirstIndexColumnCardinality();
+            int numConstants = firstColumnIteratedValues;
+            // Allow up to 30x the maximum number of probe values in IndexPrefixIteratorMode
+            // because the first column's values don't need to be compiled in the activation,
+            // so won't contribute to parsing or serialization overhead.
+            int multiplier = indexPrefixMissing ? Math.min(30, numConstants) : 1;
```

Comment:
How do we get the value of 30 here? Also, maybe consider define it as a constant.

Reply:
The idea is that, we are trying to limit both the size of the serialized operation and also the total number of values in the MultiRowRangeFilter.  The index prefix values won't contribute to the serialization size, because they are found out during execution time, but they will make the number of items in the MultiRowRangeFilter greater.  The current default limit in number of combined values is 5000.  If we had 1000 values in the first column, with no change in the code, we may end of building a MultiRowRangeFilter with 5 million items, which seems a bit high.  So the first change is to start our count at 1000 values instead of 1:

>  int numConstants = firstColumnIteratedValues;

But this means our new limit in terms of number of combined values from predicates is reduced to 5, which seems a little small.  To allow for slightly more combined values we use the multiplier 30 which would bring the limit on combined values from predicates up to 150, which results in a theoretical maximum MultiRowRangeFilter size of 150000.  Now, it is possible that filters with millions of ranges could possibly run fine on Spark.  I really haven't had the time to dive into fine tuning this and pushing the limit to find the cutoff point where large MultiRowRangeFilters start getting too expensive.  I've noticed some degradation after a couple hundred thousands ranges, and would need to try a lot more cases to feel comfortable with allowing much larger filters.  So, the value 30 comes from trying to cap the MultiRowRangeFilter at 150000 or so.  I have created a new constant for this:

> private static final int INDEX_PREFIX_MODE_MULTI_PROBE_LIMIT_MULTIPLIER = 30;

Reply:
I see. I think adding your explanation here (or the number calculations) as a comment here would be very helpful.

Reply:
OK, added this comment:

>             // Allow up to 30x the maximum number of probe values in IndexPrefixIteratorMode
>             // because the first column's values don't need to be compiled in the activation,
>             // so won't contribute to parsing or serialization overhead.
>             // The idea is that, we are trying to limit both the size of the serialized operation
>             // and also the total number of values in the MultiRowRangeFilter. The index prefix values
>             // won't contribute to the serialization size, because they are found out during execution time,
>             // but they will make the number of items in the MultiRowRangeFilter greater.
>             // The current default limit in number of combined values is 5000.
>             // If we had 1000 values in the first column, with no change in the code,
>             // we may end of building a MultiRowRangeFilter with 5 million items, which testing has shown
>             // to not perform well.  
>             // Performance starts to degrade with a MultiRowRange filter of around 150000 items,
>             // so, the value 30 for INDEX_PREFIX_MODE_MULTI_PROBE_LIMIT_MULTIPLIER comes from trying to
>             // cap the MultiRowRangeFilter at 150000 (maxMulticolumnProbeValues (5000) * 30).